### PR TITLE
[Jetpack Content Migration Flow] Add Help Screen

### LINKED ISSUE #19617: iOS - Welcome screen: Add Migration Help Screen
This task is to create a screen for the UI when users tap the "Need Help" button on the Welcome screen at the start of the migration flow during Jetpack first launch.

The Help Screen in the migration flow is in principle a trimmed down version of the Help / Support page that already exists in the app. Inspiration and possible reusable implementation can be found in the existing app screen.

The main indication for its design in this scenario is to tailor the content to be more focused (internal ref: p1667590486691019/1667578704.552489-slack-C046L7LPV0F)

Image

### PR DESCRIPTION
**Pull request by salimbraksa**
Fixes https://github.com/wordpress-mobile/WordPress-iOS/issues/19617

## Description
Display support screen when the user taps "Need help?" or "Me" avatar button in the Migration Welcome screen. This support screen is the same as the existing support screen except with few visual changes:

1. Add User Details view
2. Hide Logs / Debug section
3. Add Log out button

These visual changes are applied only when the support screen is accessed from the Migration flow.

## Upcoming PR
Since the Log Out logic is already implemented in `MeViewController`, this PR is re-using that logic, see [LogOutActionHandler](https://github.com/wordpress-mobile/WordPress-iOS/pull/19618/files#diff-dc8cfe74131a56cd0d59b3278a74e5063abe797ffde9e53caa176d8a882b0761). But this PR doesn't refactor `MeViewController` to use this new handler object, so the Log Out logic is duplicated for now. Will open another PR to fix this.

## Screenshots
| Support Screen during Migration Flow | Existing Support Screen outside Migration flow |
| -------------------------------------- | ----------------------------------------------- |
|  | iphone |

## Test Instructions:

**Support Screen during Migration flow**
1. Run Jetpack app and make sure you're logged in
2. Enable Migration Flow
   1. Locate `JetpackWindowManager.swift` file
   2. Change `shouldShowMigrationUI` value to `true`
   3. Build and run Jetpack app
3. The Migration Welcome screen should show up
4. Tap "Need Help?" or "Me" avatar button
5. **Expect** the support screen to show up and it looks like the 1st screenshot
6. Tap Logout, then confirm logging out
7. **Expect** the Jetpack Prologue screen to show up

**Support Screen outside Migration flow**
1. From the Jetpack Prologue screen, log in to your account again
2. Under "My Site" tab, tap "Me" avatar button in the top-right corner
3. Tap "Help & Support" row
4. **Expect** the support screen to show up and looks like the 2nd screenshot

## Regression Notes
1. Potential unintended areas of impact
It's worth testing that the existing support screen wasn't impacted. See **Test Instructions > Support Screen outside Migration flow**.

3. What I did to test those areas of impact (or what existing automated tests I relied on)
None.

4. What automated tests I added (or what prevented me from doing so)
None.

## PR submission checklist:
- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/Classes/ViewRelated/Support/LogOutActionHandler.swift
Comment (Gio2018) [inline WordPress/Classes/ViewRelated/Support/LogOutActionHandler.swift:37]: Should we use the new reverse dns notation here?

File: WordPress/Classes/ViewRelated/Support/SupportTableViewController.swift
Comment (Gio2018) [inline WordPress/Classes/ViewRelated/Support/SupportTableViewController.swift:387]: same consideration here about the strings

File: WordPress/Jetpack/Classes/System/JetpackWindowManager.swift
Comment (Gio2018) [inline WordPress/Jetpack/Classes/System/JetpackWindowManager.swift]: I am not sure we want to change this just yet, because the value of this flag depends on other conditions

**Review by Gio2018 — COMMENTED** [review]
Thanks for working on this @salimbraksa !
Left a few comments, let me know when the state is final so I can do a full review!

File: WordPress/Jetpack/Classes/System/JetpackWindowManager.swift
Comment (salimbraksa) [reply WordPress/Jetpack/Classes/System/JetpackWindowManager.swift]: Yes of course, I'll revert this back to `false`.

File: WordPress/Classes/ViewRelated/Support/LogOutActionHandler.swift
Comment (salimbraksa) [reply WordPress/Classes/ViewRelated/Support/LogOutActionHandler.swift:37]: Since the log out logic is already implemented in `MeViewController`, I re-used the same logic for the support screen, so those strings aren't new. I think if we change the keys, then translators will have to provide a new translations for those strings, right? If so, is there way to change the keys without losing the existing translation?

File: WordPress/Classes/ViewRelated/Support/SupportTableViewController.swift
Comment (salimbraksa) [reply WordPress/Classes/ViewRelated/Support/SupportTableViewController.swift:387]: Same as [this reply](https://github.com/wordpress-mobile/WordPress-iOS/pull/19618#discussion_r1026421774)

File: WordPress/Classes/ViewRelated/Support/SupportTableViewController.swift
Comment (salimbraksa) [inline WordPress/Classes/ViewRelated/Support/SupportTableViewController.swift:127]: Ideally, the `MeHeaderView` should be initialized with the `config` param, but the `MeHeaderView` is used elsewhere and would require some refactoring to be done.

File: WordPress/Classes/ViewRelated/Support/SupportTableViewController.swift
Comment (salimbraksa) [inline WordPress/Classes/ViewRelated/Support/SupportTableViewController.swift:127]: Ideally, the MeHeaderView should be initialized with the config param, but the MeHeaderView is used elsewhere and would require some refactoring to be done.

**Review by twstokes — COMMENTED** [review]
This is looking really good to me @salimbraksa! Nice work. 👏  I'll let @Gio2018 have the final say on the review since he started it.

**I tested:**
- Light / Dark mode
- a11y / Dynamic Type
- Accessing the Contact Support view
- Comparing the differences between the view from the migration flow and within the app

**One observation:**

When the Support view is accessed within the app we allow a landscape orientation. It appears that we don't allow landscape when coming from the migration flow. However, when I rotate my device the Status Bar and the Home Bar still rotate. This is minor, but I wanted to bring it up in case we wanted to address it.

File: WordPress/Classes/ViewRelated/Support/LogOutActionHandler.swift
Comment (Gio2018) [reply WordPress/Classes/ViewRelated/Support/LogOutActionHandler.swift:37]: I see. It's fine for me to keep them as they are, just pinging @AliSoftware since I saw a [recent issue](https://github.com/wordpress-mobile/WordPress-iOS/pull/19606https://github.com/wordpress-mobile/WordPress-iOS/pull/19606https://github.com/wordpress-mobile/WordPress-iOS/pull/19606) with localized strings, to make sure this won't cause any

Comment (AliSoftware) [reply WordPress/Classes/ViewRelated/Support/LogOutActionHandler.swift:37]: Thanks for the ping!

Using the exact same strings as the `MeViewController` is fine… **as long as you are really sure that they are used in the same kind of UI context** (and that they should thus have the exact same translations in both places, because they really are about the exact same thing)
 - That being said, instead of copy/pasting the strings and comments in there, and thus duplicating the declaration of those `NSLocalizedStrings` (between the ones in the `MeViewController` and the ones here) — which could lead to a later PR changing the comment here, but not in the other place, and lead to more confusion / discrepancies by not having a single source of truth for those code declarations), I'd instead suggest to reference the existing constants from `MeViewController` (or move those constants into a common class / file / ViewModel / what have you, as long as you declare them only once and reference them from both places)
 - This might also be a more "dangerous" assumption to make for shorter copies (like "Cancel" and "Log Out") which seems like words that are small and common enough to be used in many other places in the app. So if another place in the app has `NSLocalizedString("Cancel", comment: "Some completely different context")`, maybe the translators would be tempted to translate it differently for one context than for the other, but if you use the same key for too many different contexts and places in the code, that might lead to translation inconsistencies, [like what happened here in the past](https://github.com/wordpress-mobile/WordPress-iOS/pull/19136).

The last point and example is one more key reason why semantic keys are helpful and that it could still be a good idea to switch to them here 😉 

> If so, is there way to change the keys without losing the existing translation?

Not automatically… but almost.

It happens that GlotPress has a system they call "translation memory", and when it detects a source (English) copy for which they already had a translation for in the past, even if for a different key,… then it is automatically suggested to translators:

image

_See how this recent key in GlotPress is not translated yet… but the "Suggestions from Translation Memory" section already have translations for it, one click away to be used via the "Copy" buttons_

This means that, yes, if you change the key in this `NSLocalizedString` call to start using semantic keys here instead, the translation for that new key will start as empty / untranslated the first time that new key will be imported into GlotPress… but then it will be only one click away for polyglots to translate (picking if from the suggested translation from Translation Memory) as opposed to having to re-translate it from scratch. So that new key will get the existing translation from the old key pretty quickly, so I'd say in the end that's not that big a deal.

File: WordPress/Classes/ViewRelated/Me/Views/Header/MeHeaderView.swift
Comment (Gio2018) [inline WordPress/Classes/ViewRelated/Me/Views/Header/MeHeaderView.swift]: Likely a matter of personal preference, so feel free to ignore this comment, but I find it a bit confusing to have a nested typer that then has an extension outside the containing type. Maybe we could move `Configuration` outside `MeHeaderView`? or just add the initializer in the type (or initializers, if we want to preserve the init with all the params)?

File: WordPress/Classes/ViewRelated/Support/SupportTableViewController+Configuration.swift
Comment (Gio2018) [inline WordPress/Classes/ViewRelated/Support/SupportTableViewController+Configuration.swift]: same considerations to `Configuration` in `MeHeaderView` apply here

**Review by Gio2018 — COMMENTED** [review]
This works great @salimbraksa !

Left a few code comments; other than that I agree with what @twstokes  observes:

> When the Support view is accessed within the app we allow a landscape orientation. It appears that we don't allow landscape when coming from the migration flow. However, when I rotate my device the Status Bar and the Home Bar still rotate. This is minor, but I wanted to bring it up in case we wanted to address it.

Particularly, I agree that it's minor, but we can probably do a small change to fix the status bar.
This could be something like this:

- In `MigrationViewControllerFactory` we define a subclass of `UINavigationController`
- We override the `supportedInterfaceOrientations` property of this subclass, to only support `.portrait` on iPhone
- inside `makeSupportViewControllerRouter` we use the above subclass instead of `UINavigationController`
wdyt?

File: WordPress/Classes/ViewRelated/Me/Views/Header/MeHeaderView.swift
Comment (salimbraksa) [reply WordPress/Classes/ViewRelated/Me/Views/Header/MeHeaderView.swift]: Regarding the nested type, it is inspired from some UIKit types, like `UIButton.Configuration`, but there are other types that aren't nested, like `UINavigationBarAppearance`, so I guess it's a matter of personal preference.

Would like to hear what @twstokes thinks about this and if he prefers `Configuration` to be outside `MeHeaderView` then we make it a convention moving forward.

File: WordPress/Classes/ViewRelated/Support/LogOutActionHandler.swift
Comment (salimbraksa) [reply WordPress/Classes/ViewRelated/Support/LogOutActionHandler.swift:37]: Thanks for clarifying @AliSoftware 🙇 

>I'd instead suggest to reference the existing constants from MeViewController (or move those constants into a common class / file / ViewModel / what have you, as long as you declare them only once and reference them from both places)

I totally agree with this and the log out logic duplication is temporary. Right now the type `LogOutActionHandler` is used only in `SupportTableViewController` but I'll open another PR so that `MeViewController` uses it as well. As a result, those strings declaration will no longer be duplicated.

**Comment by salimbraksa** [comment]
Thanks @twstokes @Gio2018 for your feedback and catching the orientation issue. This commit https://github.com/wordpress-mobile/WordPress-iOS/pull/19618/commits/69738d047039ded205e183e29d8d785fc35ef4a3 resolves the orientation issue by tweaking the `preferredInterfaceOrientationForPresentation` and `supportedInterfaceOrientations` implementations in `MigrationNavigationController`.

The idea is that `MigrationNavigationController` is only responsible for forcing the `portrait` orientation **on its children**, but presented view controllers should decide their orientation.

File: WordPress/Classes/ViewRelated/Me/Views/Header/MeHeaderView.swift
Comment (twstokes) [reply WordPress/Classes/ViewRelated/Me/Views/Header/MeHeaderView.swift]: 👋 I personally don't mind nesting types that are tightly coupled with a parent type (e.g. Error, Configuration), though I recall the defaults of [SwiftLint](https://github.com/realm/SwiftLint) complaining and [there being discussions](https://stackoverflow.com/questions/42996275/nested-types-in-swift-what-is-the-good-practice) as to why.

What jumps out to me most is 

`extension MeHeaderView.Configuration {` and  `extension SupportTableViewController.Configuration {`. 

Since they have the same scope and are in the same file I'd bundle them with the type definition like @Gio2018 proposes:

> or just add the initializer in the type (or initializers, if we want to preserve the init with all the params)?

> then we make it a convention moving forward.

That'd be great to land on one!

**Review by twstokes — APPROVED** [review]
I tested the changes from 69738d047039ded205e183e29d8d785fc35ef4a3 by rotating the Support view and it worked great. The Support view _supported_ landscape and upon dismissal the view smoothly transitioned back to portrait for the migration flow.

File: WordPress/Classes/ViewRelated/Me/Views/Header/MeHeaderView.swift
Comment (salimbraksa) [reply WordPress/Classes/ViewRelated/Me/Views/Header/MeHeaderView.swift]: The only reason behind the  `MeHeaderView.Configuration` is add a convenience init without removing the default one the compiler generates. [Here is an article about adding convenience inits to structs.](https://www.swiftbysundell.com/tips/struct-convenience-initializers/)

That being said, I pushed a commit https://github.com/wordpress-mobile/WordPress-iOS/pull/19618/commits/4417b15780609f1f3da7a8a0d00bbcd04bc55791 adding the following changes:
1. Moved the `Configuration` type outside its parent type. For both `MeHeaderView` and `SupportTableViewController`. The types are named `MeHeaderViewConfiguration` and `SupportTableViewController` respectively. 
2. I kept the convenience `init(account: WPAccount)` in `extension MeHeaderViewConfiguration` so that the default init isn't removed.
3. I bundled `static func currentAccountConfiguration()` with `SupportTableViewControllerConfiguration`. Previously it was in a separate extension.

Comment (twstokes) [reply WordPress/Classes/ViewRelated/Me/Views/Header/MeHeaderView.swift]: > so that the default init isn't removed.

These changes make sense, but I wanted to ask: Is the default init needed somewhere in the codebase currently?

Comment (salimbraksa) [reply WordPress/Classes/ViewRelated/Me/Views/Header/MeHeaderView.swift]: For now, the default init is only used by the convenience init `init(account: WPAccount)`. Having the default init can also help us mock `MeHeaderViewConfiguration` in case we need to write unit tests.

**Review by Gio2018 — COMMENTED** [review]
This works as described, but, because we are only supporting portrait on iPhone in the migration flow, the animations between landscape and portrait feel a bit odd, particularly when we go back to the welcome screen (see animation below)
I think it would be preferable to just not support landscape at all, but since this could just be my personal feeling, I'll let @shaunandrews  and @osullivanchris  chime in for this.

<p align=center>

</p>

**Comment by osullivanchris** [comment]
> This works as described, but, because we are only supporting portrait on iPhone in the migration flow, the animations between landscape and portrait feel a bit odd, particularly when we go back to the welcome screen (see animation below)
I think it would be preferable to just not support landscape at all, but since this could just be my personal feeling, I'll let @shaunandrews and @osullivanchris chime in for this.

I see what you mean. It looks a bit funky in this case. But, maybe someone has a valid reason to type in landscape mode. Maybe when they want to write to support, or if they're on iPad. I think its overkill to add a portrait lock to those views. When they go back to the welcome screen its interesting that it displays in landscape for a moment then realising its locked to portrait. If it could immediately be in portrait mode that might be better. But not worth spending lots of time on.

**Comment by salimbraksa** [comment]
>...maybe someone has a valid reason to type in landscape mode. Maybe when they want to write to support.

That's a good point 👍 

>...If it could immediately be in portrait mode that might be better.

I can look into this and see if it's easily doable. But I guess it's not blocking this PR, right? @Gio2018

**Comment by twstokes** [comment]
> But, maybe someone has a valid reason to type in landscape mode.

I second this, and thought it more as a feature since the Support view already supports landscape well.

**Comment by Gio2018** [comment]
Thanks for reviewing this @osullivanchris !!

A couple of additional thoughts

> or if they're on iPad

in this case, we are not restricting to portrait only. The main reasoning behind avoiding landscape on iPhone was because most of the content of the migration views was not going to be visible and required to scroll. But then again, I can see both sides

**Comment by Gio2018** [comment]
> > ...maybe someone has a valid reason to type in landscape mode. Maybe when they want to write to support.
> 
> That's a good point 👍
> 
> > ...If it could immediately be in portrait mode that might be better.
> 
> I can look into this and see if it's easily doable. But I guess it's not blocking this PR, right? @Gio2018

yes not a blocker at all!

**Comment by osullivanchris** [comment]
Yep no blockers here for me. Whether we leave it as is, lock it to portrait (on iPhone), or do the small tweak that I suggested, I don't think the risk is high here. I'd prefer to ship as is rather than spend more than an hour on any fix 😄

**Pull request merged**
